### PR TITLE
[loc] IMU bearing correction

### PR DESCRIPTION
**Pull request by rbridges12**
Bearing correction using GPS bearing when driving in a straight line, gets published to `/corrected_bearing` in a `CorrectedBearing` struct every time odometry is published. Please add commits to display corrected bearing instead of target bearing on the GUI arrow (corrected bearing value is `CorrectedBearing.corrected_bearing`). also updated a few things in the IMU readme

### CODE REVIEW

File: jetson/filter/src/__main__.py
Comment (niwhsa9) [inline jetson/filter/src/__main__.py:198]: escape char?

File: jetson/filter/src/__main__.py
Comment (niwhsa9) [inline jetson/filter/src/__main__.py:153]: this is troublesome. Not sure if u can just set this to 0.

File: jetson/filter/src/__main__.py
Comment (niwhsa9) [inline jetson/filter/src/__main__.py:200]: im worried about the state keeping being part of loc code instead of nav telling you the last requested bearing plus last requested time.

File: jetson/filter/src/__main__.py
Comment (niwhsa9) [inline jetson/filter/src/__main__.py:347]: imu bearing and gps bearing are same frame and same limit?? Are you sure you can do this?

File: jetson/filter/src/__main__.py
Comment (niwhsa9) [inline jetson/filter/src/__main__.py:340]: don't you want this to go the other way, when the error is too high correct?

**Review by niwhsa9 — CHANGES_REQUESTED** [review]

File: jetson/filter/src/__main__.py
Comment (rbridges12) [reply jetson/filter/src/__main__.py:198]: thats just from the python autoformatter but I'll get rid of it

Comment (niwhsa9) [reply jetson/filter/src/__main__.py:198]: oh well ok thats fine then, if linter is happy im happy.

File: jetson/filter/src/__main__.py
Comment (rbridges12) [reply jetson/filter/src/__main__.py:153]: I actually think this should be fine. Since we have to be driving in order to correct the bearing, and I assume we will always have a target bearing if we are driving somewhere, then these initial values should never actually be used to check the correction condition. ig its possible that the very first iteration of nav could cause trouble, like if weve only received a single target bearing so the previous is still the default zero, but even then I doubt that all the other conditions would pass on the very first iteration of nav driving. We'll find out for sure in testing ig. Also FYI this is currently not even using the previous target bearing or rate of target bearing, we're gonna test it without that condition check first.

File: jetson/filter/src/__main__.py
Comment (rbridges12) [reply jetson/filter/src/__main__.py:200]: yeah ig there might be a problem with time inaccuracy due to network latency and LCM packet loss. again, I'm not sure if we're even gonna need these metrics for the correction, do you think its worth adding this to nav before we even test it?

File: jetson/filter/src/__main__.py
Comment (rbridges12) [reply jetson/filter/src/__main__.py:347]: yeah this is definitely wrong because I'm never modding by 360, let me think about how to fix this

File: jetson/filter/src/__main__.py
Comment (rbridges12) [reply jetson/filter/src/__main__.py:340]: no because this isnt error between actual bearing and measured bearing, this is error between target bearing and measured bearing. So if we're driving straight, nav will always line up the rover so that the error between target and measured is close to zero, thats what our pid loop does. This is just another check to make sure we're driving straight and not turning, ig its kinda redundant now that I've added the drive velocity command check. The thing I *did* get backwards was the check on rate of target bearing change which I will fix

File: jetson/filter/src/__main__.py
Comment (niwhsa9) [reply jetson/filter/src/__main__.py:153]: ok, as long as we are sure that this won't get triggered on the first iter thats fine, i guess we'd just add a flag for first iteration if it did. Like you said - should be obvious.

File: jetson/filter/src/__main__.py
Comment (niwhsa9) [reply jetson/filter/src/__main__.py:340]: oh right, my bad.

File: jetson/filter/src/__main__.py
Comment (qhdwight) [reply jetson/filter/src/__main__.py:153]: Usually prefer to use `None` in Python when a variable has "no data" and then handle it appropriately later with if statements

Comment (rbridges12) [reply jetson/filter/src/__main__.py:153]: yeah I'll do that, I was being lazy

File: config/filter/config.json
Comment (cgiger00) [inline config/filter/config.json:23]: are these the approx coordinates of ann arbor rn?

File: jetson/filter/src/__main__.py
Comment (cgiger00) [inline jetson/filter/src/__main__.py:185]: ig this might break soon, i think i just merged a pr that makes it vel_m_s instead of vel_percent

File: jetson/filter/src/__main__.py
Comment (cgiger00) [inline jetson/filter/src/__main__.py:194]: readability: self.current_target_bearing instead?

File: jetson/filter/src/__main__.py
Comment (cgiger00) [reply jetson/filter/src/__main__.py:200]: i think it'll be easier to assess after we do the long time test saturday so probably fine not to add it in before, your call tho

File: jetson/filter/src/__main__.py
Comment (cgiger00) [inline jetson/filter/src/__main__.py:130]: these are no longer correct right?

File: jetson/filter/src/__main__.py
Comment (cgiger00) [inline jetson/filter/src/__main__.py:265]: is it backwards on the gui that the gps coordinates are north & west?

File: jetson/filter/src/__main__.py
Comment (cgiger00) [inline jetson/filter/src/__main__.py:334]: no longer exists?

File: jetson/filter/src/__main__.py
Comment (cgiger00) [inline jetson/filter/src/__main__.py:363]: comment as a (maybe) todo to add back in?

File: jetson/imu_channel/README.md
Comment (cgiger00) [inline jetson/imu_channel/README.md:71]: nice

File: jetson/imu_channel/README.md
Comment (cgiger00) [inline jetson/imu_channel/README.md:106]: 'add in bearing correction if necessary' or smth of the sort

Comment (cgiger00) [inline jetson/imu_channel/README.md:106]: would be very useful! could be modified to work for the spectral triad too ☠️

File: jetson/filter/src/__main__.py
Comment (cgiger00) [inline jetson/filter/src/__main__.py:481]: i guess to save some space you could use the same Odometry object but publish the corrected_bearing to this one's bearing and put it to "/corrected_bearing", gives extra info but i think for testing it's an easy add on so we don't need to delete the CorrectedBearing lcm after we're done testing/leave a random unused one hanging

Comment (cgiger00) [reply jetson/filter/src/__main__.py:481]: would put extra unnecessary overhead onto the network tho but i dont think it's a big deal for testing

File: jetson/filter/src/__main__.py
Comment (qhdwight) [reply jetson/filter/src/__main__.py:130]: Here are the correct status names:

```c++
// Gets the string representation of a nav state.
std::string StateMachine::stringifyNavState() const {
    static const std::unordered_map<NavState, std::string> navStateNames =
            {
                    {NavState::Off,                  "Off"},
                    {NavState::Done,                 "Done"},
                    {NavState::DriveWaypoints,       "Drive Waypoints"},
                    {NavState::BeginSearch,          "Change Search Algorithm"},
                    {NavState::Search,               "Search"},
                    {NavState::DriveToTarget,        "Drive to Target"},
                    {NavState::TurnAroundObs,        "Turn Around Obstacle"},
                    {NavState::DriveAroundObs,       "Drive Around Obstacle"},
                    {NavState::SearchTurnAroundObs,  "Search Turn Around Obstacle"},
                    {NavState::SearchDriveAroundObs, "Search Drive Around Obstacle"},
                    {NavState::BeginGateSearch,      "Gate Prepare"},
                    {NavState::GateTraverse,         "Gate Traverse"},

                    {NavState::Unknown,              "Unknown"}
            };

    return navStateNames.at(mRover->currentState());
} // stringifyNavState()
```

Comment (qhdwight) [reply jetson/filter/src/__main__.py:130]: I'm now noticing `BeginSearch` is named wrong

File: config/filter/config.json
Comment (rbridges12) [reply config/filter/config.json:23]: yep, ig a sidenote is we need to remember to change this for URC and CIRC

File: jetson/filter/src/__main__.py
Comment (rbridges12) [reply jetson/filter/src/__main__.py:185]: yeah ill merge that into my branch before this gets merged

File: jetson/filter/src/__main__.py
Comment (rbridges12) [reply jetson/filter/src/__main__.py:130]: that list of static nav states is left over from the KF, it isn't actually used at all rn. on that note, should I just remove all the KF code since we aren't gonna be using it?

File: jetson/imu_channel/README.md
Comment (rbridges12) [reply jetson/imu_channel/README.md:106]: yeah thats exactly what I was thinking, should be like 2 lines long if arduino CLI and all the necessary libs are already installed on the base station

File: jetson/filter/src/__main__.py
Comment (rbridges12) [reply jetson/filter/src/__main__.py:481]: I mean either way we still have to remember to delete something after testing it, I think its gonna be easier to just leave it like this and ill be sure to get rid of that LCM struct once were done testing this

File: config/filter/config.json
Comment (cgiger00) [reply config/filter/config.json:23]: what should they be changed to for each respective comp?

File: jetson/filter/src/__main__.py
Comment (cgiger00) [reply jetson/filter/src/__main__.py:130]: honestly, yeah, if you chose to rewrite it you can always use the git history as a place to start, or start w smth new

File: config/filter/config.json
Comment (rbridges12) [reply config/filter/config.json:23]: I think any GPS point within a few miles of the comp should be fine

File: jetson/filter/src/__main__.py
Comment (niwhsa9) [reply jetson/filter/src/__main__.py:192]: current is redundant

File: jetson/filter/src/inputs.py
Comment (cgiger00) [inline jetson/filter/src/inputs.py:255]: can this be a constant or commented somewhere as the invalid value?

**Pull request closed without merging**
